[export] Refactor `constrain_as_value` and `constrain_as_size`

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106591

Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

tugsbayasgalan committed:
[export] Refactor constrain_as_value and constrain_as_size

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Refactor constrain_as_value and constrain_as_size"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Refactor constrain_as_value and constrain_as_size"


Some notable changes:
1. `constrain_as_size` now only takes `max` value as input. This is because we want to better distinguish between `constrain_as_value` and `constrain_as_size`. This is technically BC breaking, but like, no one is using this API rn. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -183,12 +183,11 @@
   dispatch:
-    CPU: sym_constrain_range_cpu
-    CUDA: sym_constrain_range_cuda
+    CompositeExplicitAutograd: sym_constrain_range
```
I think so lol

ezyang commented on test/dynamo/test_export.py:
```diff
@@ -2321,7 +2321,7 @@ def test_exported_graph_serialization(self):
         def f(x, y):
             b = x.item()
-            constrain_as_size(b, min=2, max=5)
+            constrain_as_size(b, max=5)
```
oh, you didn't end up deleting min/max from constrain_as_size?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -346,13 +367,9 @@ def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
         values at runtime (we assume that a graph that is valid for N=2 will
         also be valid for N=1).
     """
+    assert max is not None
```
How come? 🤔

ezyang commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,96 @@
     """
+    # NOTE: If min value is not passed, we will assume it means there is no
+    # user specified min value provided. If this happens, we will assume min value to
+    # be 2 for compiler only! Runtime will assume min value will be 0.
```
But there's no int anymore

ezyang commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,96 @@
-        )
-    return constrain_as_value(symbol, min, max)
+    assert isinstance(symbol, SymInt)
+    return torch.sym_constrain_range(symbol, min=None, max=max)
```
Doesn't this expect None max?

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` now only takes `max` value as input. This is because we want to better distinguish between `constrain_as_value` and `constrain_as_size`. This is technically BC breaking, but like, no one is using this API rn. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,32 @@
-    return constrain_as_value(symbol, min, max)
+    # NOTE: If min, max value are not passed, we will assume it means this is only used for compiler hint.
+    # Runtime will assume min value will be 0 and max value will be INT_MAX
+    torch.sym_constrain_range(symbol, min=None, max=None)
```
This doesn't seem like it will work? I feel like you need two ops.

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` doesn't take any min/max value. This is because we want to better distinguish between `constrain_as_value` and `constrain_as_size`. This is technically BC breaking, but like, no one is using this API rn. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -181,14 +181,21 @@
-    CUDA: sym_constrain_range_cuda
+    CompositeExplicitAutograd: sym_constrain_range
+
+- func: sym_constrain_range_for_size(Scalar size) -> ()
```
This naming is weird because there's no range anymore. Why not just call it the public API name (`constrain_as_size`) (I'm also super unclear about why this operators have `sym_` prefix on their name, it seems unnecessary)

ezyang commented on test/dynamo/test_export.py:
```diff
@@ -2321,7 +2321,7 @@ def test_exported_graph_serialization(self):
             b = x.item()
-            constrain_as_size(b, min=2, max=5)
+            constrain_as_size(b)
             return torch.empty((b, y.shape[0]))
```
🤔  Maybe we should automatically constrain as size whenever you pass data dependent values in sizey contexts. This would be consistent with the "special irrefutable assert" construct we've been talking about.

ezyang replied:
(no action needed)

gmagogsfm replied:
Yeah, not bad, we compromise a big of explicit-ness but users get less hassle in most cases.

gmagogsfm replied:
s/big/bit/

ezyang commented on torch/_custom_op/impl.py:
```diff
@@ -766,7 +766,7 @@ def create_unbacked_symint(self, *, min=2, max=None) -> torch.SymInt:
 
         result = self._shape_env.create_unbacked_symint()
-        torch.fx.experimental.symbolic_shapes.constrain_range(result, min=2, max=max)
+        torch.fx.experimental.symbolic_shapes.constrain_range(result, min=2, max=max if max else sys.maxsize)
```
Is this change required?

I have a good reason to not want to pass sys.maxsize as the upper bound: it will result in extra arbitrary precision integer math in many situations. For example, suppose I have x bounded (2, inf), if I multiple x by two its modified bound is (4, inf) (easy), but if I had maxsize, the modified bound is (4, maxsize * 2) (which is only representable in arbitrary precision.)

tugsbayasgalan replied:
Ohh this is not needed anymore. It was left from previous iteration.

ezyang commented on test/export/test_export.py:
```diff
@@ -37,23 +37,6 @@ def f(x):
-
-        inp = (torch.tensor([6]),)
-        with self.assertRaisesRegex(torchdynamo.exc.UserError, "Invalid value 6 for range"):
-            export(invalid_input_conflict_with_inline_constraints, inp)
```
You can still do this test, so long as you actually run the resulting exported module (it should error at runtime)

ezyang commented on test/export/test_export.py:
```diff
@@ -72,8 +55,8 @@ def invalid_input_conflict_with_input_constraints(x):
-            constrain_as_size(b, min=2, max=3)
-            constrain_as_size(b, min=4, max=5)
+            constrain_as_size(b)
+            constrain_as_value(b, min=4, max=5)
```
This isn't really conflicting constraints anymore, right? Why is it erroring?

ezyang commented on aten/src/ATen/native/Constraints.cpp:
```diff
@@ -20,16 +21,41 @@
+    c10::optional<int64_t> max) {
 
-Tensor _functional_sym_constrain_range_cpu(
+    int64_t min_val = min.has_value() ? min.value() : 0;
```
This is flat out wrong. If min is not specified, the minimum is unbounded below (could be negative). You need to simply skip the le check if it is none. Ditto below (where it is correct, but it will be more symmetric to skip the upper bound test when it's none.)

ezyang commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,32 @@
+    if min is None:
+        min = 0
+    if max is None:
+        max = sys.maxsize
```
I still don't think you should do this. What's the point of the op taking optional arguments if you always fill them in?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -312,6 +312,39 @@ def _constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
+
+def _constrain_unbacked_symint_for_size(a):
+    if isinstance(a, int):
+        return
```
This is weirdly asymmetric with `_constrain_symbol_range`. Don't you want to assume you're getting a sympy.Symbol as argument?

ezyang replied:
(and then the enclosing code needs to do a >= 0 test if you do get an int)

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -312,6 +312,39 @@ def _constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
+            builtins.max(r.lower, 2), builtins.min(r.upper, sys.maxsize)
+        )
+    else:
+        shape_env.var_to_range[s] = ValueRanges(2, sys.maxsize)
```
Don't use maxsize, as above

tugsbayasgalan commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -312,6 +312,39 @@ def _constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
     else:
         shape_env.var_to_range[s] = ValueRanges(min, max)
 
+    shape_env.runtime_var_to_range[s] = shape_env.var_to_range[s]
```
Hmm my thinking was if the same symbol that was used for `constrain_as_size` is used again to be constrained as value, we should respect the latter option.

ezyang replied:
You do need to update runtime_var_to_range, but it shouldn't be an overwrite (an intersection, I think?)

tugsbayasgalan replied:
Yeah thats true.

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` doesn't take any min/max value. This is because we want to better distinguish between `constrain_as_value` and `constrain_as_size`. This is technically BC breaking, but like, no one is using this API rn. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` doesn't take any min/max value. This is because we want to better distinguish between `constrain_as_value` and `constrain_as_size`. This is technically BC breaking, but like, no one is using this API rn. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on test/test_proxy_tensor.py:
```diff
@@ -1128,7 +1130,7 @@ def f(x, mask, params, buffers):
                     guard_int(s)
             x = x[mask]
-            constrain_range(x.shape[0], min=1)
+            constrain_as_value(x.shape[0], min=1)
```
Does `constrain_as_value` show up in the graph too? Hard to tell here.

tugsbayasgalan replied:
It will show up as aten.sym_constrain_range

ezyang commented on aten/src/ATen/native/Constraints.cpp:
```diff
@@ -15,21 +16,48 @@
+  return dep_token.clone();
+}
+
+void sym_constrain_for_size(const Scalar& size) {
```
You need to test that it's an int, and that it's >= 0, no?

ezyang commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,25 @@
-    return constrain_as_value(symbol, min, max)
+    # NOTE: If min, max value are not passed, we will assume it means this is only used for compiler hint.
+    # Runtime will assume min value will be 0 and max value will be INT_MAX
+    torch.sym_constrain_for_size(symbol)
```
Why do we not line up the internal aten name with the public name here 🤔

ezyang commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,25 @@
     """
-    Add min/max constraint on the intermediate symbol which will be used as a size
+    Add compiler hints to the intermediate symbol which will be used as a size
+    in another tensor.
```
Here's the docblock I would write:

This indicates that a given int is size-like, and can be used in any context where a size is expected. You will typically use this when reading out unbacked integers from Tensors, e.g., max.item() or lengths.tolist() which then need to be used as tensor constructors. Providing these assertions to PyTorch can help resolve GuardOnDataDependentSymNode errors upon export, since we cannot guard on unbacked SymInts.

This function has unusual semantics which distinguish it from constrain_as_value. Specifically, at compile-time, we will unsoundly assume that the resulting int is always `>= 2`. This makes it easier to use the unbacked int in size contexts, as we will often attempt to guard on a size being zero/one (e.g., when computing the contiguity of a tensor, or testing if broadcasting can occur), which will not work on unbacked SymInts. Assuming that the int is `>= 2` allows us to report False to these tests. Although this is technically unsound, in practice we observe that if your program works for all sizes `>= 2`, it probably works for zero and one too. At runtime, we only assert that the size-like int is `>= 0`, matching this intuition. (This is in contrast to `constrain_as_value(s0, min=2)` which would require, at runtime, that s0 is >= 2, or `constrain_as_value(s0, min=0)`, which typically isn't enough information to let you use s0 as a size to construct tensors.)

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` doesn't take any min/max value. This is because we want to better distinguish between `constrain_as_value` and `constrain_as_size`. This is technically BC breaking, but like, no one is using this API rn. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

gmagogsfm commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,59 @@
     """
-    Add min/max constraint on the intermediate symbol which will be used as a size
+    This indicates that a given int is size-like, and can be used in any context where a size is expected.
+    You will typically use this when reading out unbacked integers from Tensors, e.g., max.item() or lengths.tolist()
```
```suggestion
    You will typically use this when reading out integers from Tensors, e.g., max.item() or lengths.tolist()
```

gmagogsfm replied:
users won't know what "unbacked" is

gmagogsfm commented on torch/_export/constraints.py:
```diff
@@ -1,48 +1,59 @@
+      GuardOnDataDependentSymNode errors upon export, since we cannot guard on unbacked SymInts.
+
+    This function has unusual semantics which distinguish it from constrain_as_value.
+    Specifically, at compile-time, we will unsoundly assume that the resulting int is always >= 2.
```
Maybe mention the reason why we always assume it to be >= 2?

gmagogsfm commented on torch/_export/__init__.py:
```diff
@@ -398,6 +398,8 @@ def convert_to_fake(x):
 
         except (ConstraintViolationError, ValueRangeError) as e:
             raise UserError(UserErrorType.CONSTRAIN_VIOLATION, str(e))
+        except torch._dynamo.exc.TorchRuntimeError as e:
```
All all runtime errors because of runtime assertion failures due to breaking constraints?

tugsbayasgalan replied:
I guess it is ok to directly raise TorchRuntimeError.

gmagogsfm commented on torch/_meta_registrations.py:
```diff
@@ -407,6 +412,19 @@ def functional_sym_constrain_range(size, min, max, dep_token):
+@register_meta(aten._functional_sym_constrain_range_for_size.default)
+def functional_sym_constrain_range_for_size(size, min, max, dep_token):
+    aten.sym_constrain_range_for_size(size, min=min, max=max)
+    return dep_token
```
Do we need to clone dep_token instead of returning itself directly?

tugsbayasgalan replied:
I think that is only true for Cpp kernels.

ezyang replied:
No, you need to clone here too lol

gmagogsfm commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -304,13 +304,57 @@ def guard_scalar(a):
+        a.node.shape_env,
+        a.node.expr,
+        compiler_min=compiler_min,
+        compiler_max=max,
```
This `compiler_max` isn't really needed, are you doing it for symmetry?

tugsbayasgalan replied:
Yeah just for symmetry.

tugsbayasgalan commented on test/dynamo/test_export.py:
```diff
@@ -2273,7 +2273,7 @@ def my_dyn_fn(a, b, c):
         def f(x, y):
             b = x.item()
-            constrain_as_size(b, min=2, max=5)
+            constrain_as_size(b)
```
We do. I have some other tests that use min/max value. I was too lazy to change this back lol.

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -f "Not related failures"

izaitsevfb commented:
@pytorchbot revert -m "Breaks inductor test on trunk" -c nosignal

see: https://github.com/pytorch/pytorch/actions/runs/5829257494/job/15808815449

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan committed:
Update on "[export] Refactor `constrain_as_value` and `constrain_as_size`"


Some notable changes:
1. `constrain_as_size` allows min value to be less than 2 as it will unconditionally assume min >= 2 for compiler purposes. Instead, we add additional check to make sure max value is always greater than 2. 
2. Previously, we used to runtime assert on the unbacked symint's val range which would be always between [2, max]. I modified this logic to assert on [0, max] unless user explicitly specifies the min range. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge
